### PR TITLE
DTLS fragments as DTLSMessageHandshakeBody::Fragment

### PR DESCRIPTION
**Pull request by algesten**
TL/DR: To combine fragments into a full message some fields are needed that are available in the output of a parser that currently fails for fragments (`parse_dtls_message_handshake`). Those fields are not available in the output of `parse_dtls_raw_record`.

--------

DTLS fragment parsing is currently harder than it ought to be. Attempting to parse a fragment with [`parse_dtls_message_handshake`](https://github.com/rusticata/tls-parser/blob/36a0022029f104e0b3cc6e9dd18ad3da85b106e0/src/dtls.rs#L219C1-L219C32) fails in an unexpected way. The problem is in this code.

https://github.com/rusticata/tls-parser/blob/36a0022029f104e0b3cc6e9dd18ad3da85b106e0/src/dtls.rs#L219-L225

The last line `take(length)(i)?;` is using `length`, which is the entire length of the _unfragmented_ message. For a fragment, that amount of data is not available in `i`, making the parsing fail on a length check (arguably a bit strange for the user). The length for the current packet is `fragment_length`, which is `== length` for non-fragmented packets.

That means the current `parse_dtls_message_handshake` can only handle _complete_ messages, no fragments, which means the user must manually combine multiple messages together to do a successful parse.

For this purpose we have [`parse_dtls_raw_record`](https://github.com/rusticata/tls-parser/blob/36a0022029f104e0b3cc6e9dd18ad3da85b106e0/src/dtls.rs#L286C8-L286C29). This parses the TLS record header, and leaves the entire message(s) as raw data. But to be able to combine multiple fragments into the complete message, the TLS record header is not enough. We also need the `fragment_offset` and `fragment_length` that are deeper down in the handshake (see code above).

The reason is that DTLS fragments are allowed to overlap. [DTLS 1.2 RFC section 4.2.3](https://www.rfc-editor.org/rfc/rfc6347#section-4.2.3)

```text
   When a DTLS implementation receives a handshake message fragment, it
   MUST buffer it until it has the entire handshake message.  DTLS
   implementations MUST be able to handle overlapping fragment ranges.
   This allows senders to retransmit handshake messages with smaller
   fragment sizes if the PMTU estimate changes.
```

This PR suggest a (breaking) change to `parse_dtls_message_handshake` making it handle fragments in addition to what it currently does.

It extends the `DTLSMessageHandshakeBody` with a new enum value `Fragment` meaning it can be any of the other body types, but not complete yet.

This allows the `parse_dtls_message_handshake` to succeed also for fragments, and the user can get the `fragment_offset`/`fragment_length` fields necessary to make a complete message.

### CODE REVIEW

File: src/dtls.rs [hunk 1]
```diff
@@ -148,6 +160,11 @@ pub fn parse_dtls_record_header(i: &[u8]) -> IResult<&[u8], DTLSRecordHeader> {
     Ok((i, record))
 }
 
+/// Treat the entire input as an opaque fragment.
+fn parse_dtls_fragment(i: &[u8]) -> IResult<&[u8], DTLSMessageHandshakeBody> {
+    Ok((&[], DTLSMessageHandshakeBody::Fragment(i)))
+}
+
```
Comment (chifflier) [inline src/dtls.rs:167]: I understand we may not know the length of the fragment, but can't this consume more than needed? Can there be data after the fragment?

Comment (chifflier) [reply src/dtls.rs:167]: Second thought: maybe this should be bound to at most `fragment_length` (from the caller)?

**Comment by chifflier** [comment]
I understand the propose change to the API and I think this is indeed required to better handle the fragments.

Some logic will be necessary to combine them after, but it should not be done during the parsing (and it will probably be tricky, because of the possible overlaps and the multiple bounds checks required to do it properly).

I'd like to merge the PR, but would prefer to wait for answers for my comments above.
Thanks!

File: src/dtls.rs (hunk 1 above)
Comment (algesten) [reply src/dtls.rs:167]: If you check where this is called from, I think your concern is addressed.

```rs
 let (i, raw_msg) = take(fragment_length)(i)?;
```

`fragment_length` read from the header, controls the amount of data we consume from the input. It won't consume too much unless the DTLS header is malformed.

Comment (chifflier) [reply src/dtls.rs:167]: Oh, I see, the function comment lead me to believe it was called with the entire output, while it is (and must be) called with raw fragment data.

Comment (algesten) [reply src/dtls.rs:167]: Can I improve the function name and/or maybe some code doc?

Comment (chifflier) [reply src/dtls.rs:167]: This should be a comment in this function indicating the preconditions. However, I must admit some other (similar) functions do not all have comments

**Comment by chifflier** [comment]
Given that everything looks addressed (except maybe a comment, but that can be added later), I've decided to merge this PR now, so I can go into more tests (like fuzzing).
Thank you for your contribution!
